### DISCUSSION TITLE
Proper Type Aliases

### DISCUSSION THREAD

**Discussion by ziaulhasanhamim** [dotnet/csharplang#5579] [Language Ideas] [answered]
In C# there is no way to create actual type aliases. The type usings are scoped in one file only. You can't import them. Type aliases can be very useful with structs and generics.

An example with a package OneOf which I use very frequently.
```c#
using OneOf;
namespace Application.Utils
{
  typealias AuthResult = OneOf<User, InvalidName, NameTaken>;
}
```
```c#
using Application.Utils;

AuthResult createUserResult = CreateUser(username);
var response = createUserResult.Match(
  user =>
  {
      // code
   },
   invalidName => 
   {
      // code 
   },
   nameTaken => 
   {
      // code
   });
```

**Comment by CyrusNajmabadi** [comment]
> The type usings are scoped in one file only.

This is not true. 

```c#
global using AuthResult = OneOf<User, InvalidName, NameTaken>;
```

**Reply by ziaulhasanhamim** [reply]
that's scoped to the project. Type Aliases should behave like types. You can only use them by importing the namespace with using statement or doing `Application.Utils.AuthResult`. You can use this alias from other projects too.

**Reply by jnm2** [reply]
You can, however, "export" them by adding `<Using>` items in a NuGet package .props file.

**Reply by ziaulhasanhamim** [reply]
@jnm2 isn't that too much for an alias? There should be a simple option. Like this: https://doc.rust-lang.org/reference/items/type-aliases.html. Type aliases might also help in type-driven logic.

```c#
typealias Kilograms = double;

public class Package 
{
    public Kilograms weight { get; set; }
}
```

Most modern language has some support for aliasing types

**Reply by jnm2** [reply]
I don't know, maybe a case could be made that this is "too much." For a language feature, I expect that type aliases aren't as useful as a hypothetical feature that would stop you from accidentally passing `Kilograms` where `Kilometers` is expected.

**Reply by ziaulhasanhamim** [reply]
Yeah but it states that. The unit of the value is in kg, not in grams. You don't have to name the property `WeightInKG`

**Reply by jnm2** [reply]
That's more reason why you'd want the compiler to do some sort of checking, since the source code doesn't look like `WeightInKG = newWeightInGrams` but rather it looks like `Weight = newWeight`. If `Weight` is `Kilograms` and `newWeight` is `Grams`, type aliases will not cause there to be a compiler warning or error. For that, you'd need a solution that is not a simple alias for `double`/`decimal`/whatever.

"Type alias" means "another name for the same concept," but you might not want kg and g to be treated by the compiler as the same concept (`double`). If you don't, then type aliases aren't the tool for the job.

**Reply by HaloFour** [reply]
> "Type alias" means "another name for the same concept," but you might not want kg and g to be treated by the compiler as the same concept (`double`).

I'm kinda hoping that there's room for that in the discussion of roles.  But I agree that type aliases aren't the solution here as they don't provide any compile-time type checking.

**Reply by TahirAhmadov** [reply]
For the physical measurements scenario, I agree with others that aliasing is the wrong approach. However, roles are also the wrong approach, IMO. A wrapper struct is in order here:
```cs
public struct Weight 
{ 
  public static Weight FromG(double g) => new Weight(g / 1000);
  // FromOz, etc.
  double _kg; 
  public Weight(double kg) { this._kg = kg; } 
  public double KG => this._kg;
  public double G => this._kg * 1000;
  // Oz, etc.
}
```

**Reply by HaloFour** [reply]
That will depend on how roles are eventually designed, but I see them serving exactly as your wrapper struct, just without the wrapper struct.

**Reply by TahirAhmadov** [reply]
The problem with roles the way they are _currently_ proposed, is that it's way too easy to "cast" between them. 
```cs
role Weight : double { }
role Length : double { }
Weight w = 10;
Length l = (double)w;
// or even
Length l = w; // I'm not sure if this is allowed under current proposal?
```
**If** that design is altered significantly, then we can talk about the possibility of using roles for this, but that would be predicting the future at this point.

**Reply by CyrusNajmabadi** [reply]
> the problem with roles the way they are currently proposed, is that it's way too easy to "cast" between them.

I don't think that's easy at all.  That's an explicit cast away from the type to another type.  If you are actively working to move in/out of the domains then i don't see the problem.  The same as if there were actual strong types here and you did something like call `Gram g = new Gram(myKg.RawValue);`   

If you choose to move things between domains, then that's on you.

**Reply by CyrusNajmabadi** [reply]
> Length l = w; // I'm not sure if this is allowed under current proposal?

It is explicitly called out as likely not being something we'd want precisely because of this problem.

**Reply by TahirAhmadov** [reply]
What about reflection? What about casting to `object` then back to the "role"? (Including `is` and `as`)
Another question: down to the metal, is there a performance hit for using wrapper structs, or it's optimized to a simple numeric assignment?

**Reply by CyrusNajmabadi** [reply]
> What about reflection? 

No idea.  It will have to be determined as part of this process with the runtime team.

> What about casting to object then back to the "role"?

No idea.  It will have to be determined as part of this design process.

> down to the metal, 

This question cannot be answered now as we do not even have confirmation on what this will compile to, or what changes the runtime may have.

The goal is certainly that there would be no overhead, but it's too early to say, and to say if that will be true for all operations/behaviors.

**Reply by TahirAhmadov** [reply]
> > down to the metal,
> 
> This question cannot be answered now as we do not even have confirmation on what this will compile to, or what changes the runtime may have.

Just to clarify - I wasn't asking about roles, but rather, struct wrappers as they work today. Between the compiler and JITter, can `var x = new Weight(2.0);` become `var x = 2.0;` in "assembly"? Or it outputs something more complicated, like some kind of stack alloc and ctor method call?

**Reply by CyrusNajmabadi** [reply]
It is up to the JIT.  The JIT is free to optimize it as much as it wants as long as it is not observable. 

Note: the representation of a struct is really just going to be the flattened representation of its individual parts.  THere's nothing else to 'allocate'.  There may be a Type somewhere floating around *if necessary* to pass type-info along.  But that's only if needed.  For example, if you boxed this into an object.  It would need that Type so that queries against that boxed value would be able to answer correctly.

**Reply by TahirAhmadov** [reply]
I asked because if the simple form (no boxing) as shown below:
```cs
struct Weight { double _value; ... }
Weight w = new(2.0);
class Class { Weight _w = new(2.0); }
```
... is actually compiled and JITted to:
```cs
double w = 2.0;
class Class { double _w = 2.0; }
```
... from memory and actual CPU operations perspective, then the currently available wrapper structs basically give you compile time safety for free, it's a very good technique to use - if developers were to have doubts about this optimization, I could see them (including myself) being uneasy about wrapping these `double`s in domain specific structs; but if it's known that the optimization basically "erases" the struct (in simple scenarios), it's one less thing to worry about.
PS. And even when you need to box the wrapper struct, it's really not much different than boxing a double. the cost comes from the allocation, there is no additional cost because `Weight` is somehow "heavier" (pun intended) than `double` to box - am I right?

**Reply by theunrepentantgeek** [reply]
Playing around with [sharplab.io](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQ6LAE4FcDGwABAOoCmAlgOYAWx6A3ukS0QCYD2uARgDZlEAbmF64yAbiLNW0lnADMpSrWAAKTj35CAlFLSsiTfQdbDRAgLxDxsogF9bthUQ6Cy2bBTYC4SAAxEACocAMo4FAB2VKq6tkYmrHAA7EIiYgB0wWGeUTE2xiwOaEXozohEAMJ68YmKcAAsRACyMXoGNQnk1HREAO5EVhFkvaoI6X7a+QlySACcqr2TtkV2QA===) is a really good way to explore this.

**Reply by CyrusNajmabadi** [reply]
> ... from memory and actual CPU operations perspective,

First, any answer is inherently difficult here as this is a point-in-time question.  What may be true today many not be true tomorrow.

That said, from standard speceeze for both C# and the CLR, we can say that from the "memory" perspective this will certainly be true.  This is effectively the definition of what a struct is.  The memory of a struct is the memory of its contents (modulo packing concerns which don't arise here).

For CPU you'd have to measure.  However, with modern jit this shouldnt' be a problem.

> if developers were to have doubts about this optimization,

Here's the thing.  Either the optimization doesn't matter to you, in which case you don't hve to care.  Or it matters, in which case you need to measure no matter what. :)

So what i say isn't really relevant here.

> but if it's known that the optimization

Try it out and see :)

> PS. And even when you need to box the wrapper struct, it's really not much different than boxing a double. the cost comes from the allocation, there is no additional cost because Weight is somehow "heavier" (pun intended) than double to box - am I right?

This is less clear.  The runtime may certainly try to optimize certain boxes and boxed values more than others.  As a (trivial) example, you might imagine the runtime optimizin boxing `true` and `false` into only two objects max (it doesn't do this, but you could imagine it).  HOwever, once the bool was wrapped into another type this might not happen.

The best thing to do with perf is measure :)

**Reply by TahirAhmadov** [reply]
> For CPU you'd have to measure. However, with modern jit this shouldnt' be a problem.

I'm wondering what would be a good forum to post this question, specifically in the context of recent "unfolding" work. Would that be the runtime repo?

> > if developers were to have doubts about this optimization,
> 
> Here's the thing. Either the optimization doesn't matter to you, in which case you don't hve to care. Or it matters, in which case you need to measure no matter what. :)

I disagree on this point. **Yes,** I want performant code. **No,** I don't want to spend time measuring everything. 
I want to remember that "approach A" gives me "result X". Yes, A can get hairy (as described below with bool true and false); but if I have clarity for 90%, and I can apply that knowledge to produce code that's 90% of perfectly performant, I'm (and my clients are) happy. 

> > PS. And even when you need to box the wrapper struct, it's really not much different than boxing a double. the cost comes from the allocation, there is no additional cost because Weight is somehow "heavier" (pun intended) than double to box - am I right?
> 
> This is less clear. The runtime may certainly try to optimize certain boxes and boxed values more than others. As a (trivial) example, you might imagine the runtime optimizin boxing `true` and `false` into only two objects max (it doesn't do this, but you could imagine it). HOwever, once the bool was wrapped into another type this might not happen.

I read somewhere (I can't remember exactly where now) that they have some optimizations for some values, and it's very empirical evidence driven. Essentially, that team looked at real life usage and said, for `int`, values like 0, 1, 10, 20, 100, (obviously I don't know the actual values - just making a point here) are common, and we'll "cache" those; and others, we'll generate new heap allocations. I think with bool it's entirely possible (it wasn't called out in that article) that it's trivial to cache `true` and `false`. However, this approach breaks apart with `double`. No way they can deduce a "set" of double values to cache (maybe with the exception of 0.0).

**Comment by fa1ntl1ght** [comment]
#5497

**Comment by MCGPPeters** [comment, marked as answer]
Take a look at https://github.com/dotnet/csharplang/discussions/5496 . 

Specifically the paragraph on "Generalized aliases"

**Reply by ziaulhasanhamim** [reply]
Yeah, that's what I was talking about.
